service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java

```diff
@@ -394,4 +427,52 @@ protected final CorsHandler createCorsHandler(final String allowedOrigin, final
                 .allowedHeader(HttpHeaders.IF_MATCH.toString())
                 .exposedHeader(HttpHeaders.ETAG.toString());
     }
+
+    /**
+     * Creates a response based on generic result.
+     * <p>
+     * The behavior is as follows:
+     * <ol>
+     * <li>Set the status code on the response.</li>
+     * <li>If the status code represents an error condition (i.e. the code is &gt;= 400),
+     * then the JSON object passed in to the returned handler is written to the response body.</li>
+     * <li>Otherwise, the JSON object is written to the response body and the given custom handler is
+     * invoked (if not {@code null}).</li>
+     * </ol>
+     *
+     * @param ctx The routing context of the request.
+     * @param result The generic result of the operation.
+     * @param customHandler An (optional) handler for post processing the HTTP response, e.g. to set any additional HTTP
+     *                      headers. The handler <em>must not</em> write to response body. May be {@code null}.
+     */
+    protected final void writeResponse(final RoutingContext ctx, final Result<?> result, final Handler<HttpServerResponse> customHandler) {
+        final int status = result.getStatus();
+        final HttpServerResponse response = ctx.response();
+        response.setStatusCode(status);
+        if (status >= 400) {
+            HttpUtils.setResponseBody(response, JsonObject.mapFrom(result.getPayload()));
+        } else if (status == HttpURLConnection.HTTP_CREATED) {
+            HttpUtils.setResponseBody(response, JsonObject.mapFrom(result.getPayload()));
+            if (customHandler != null) {
+                customHandler.handle(response);
+            }
+        }
+        response.end();
+    }
+
+    /**
+     * Creates a response based on operation result (including resource version).
+     * <p>
+     * Sets ETAG header and then calls {@link #writeResponse}
+     *
+     * @param ctx The routing context of the request.
+     * @param result The operation result of the operation.
+     * @param customHandler An (optional) handler for post processing the HTTP response, e.g. to set any additional HTTP
+     *                      headers. The handler <em>must not</em> write to response body. May be {@code null}.
+     */
+    protected final void writeOperationResponse(final RoutingContext ctx, final OperationResult<?> result, final Handler<HttpServerResponse> customHandler) {
+        result.getResourceVersion().ifPresent(v -> ctx.response().putHeader(HttpHeaders.ETAG, v));
+        writeResponse(ctx, result, customHandler);
+    }
```

Comment:
I have to admit that I do not really like the dependency on the management package that is being introduced here. Up until now, this class (and package) has been providing generic functionality for implementing HTTP endpoints. These imports now tie it to the management API which I don't think is desirable nor necessary. How about moving these two methods to an abstract endpoint class in the `org.eclipse.hono.service.management` package?

Reply:
These are now moved to the Util class.